pkg/extensions/sync/destination.go

```diff
@@ -80,178 +82,209 @@ func (registry *DestinationRegistry) CanSkipImage(repo, tag string, imageDigest
 	return true, nil
 }
 
-func (registry *DestinationRegistry) GetContext() *types.SystemContext {
-	return registry.tempStorage.GetContext()
-}
-
-func (registry *DestinationRegistry) GetImageReference(repo, reference string) (types.ImageReference, error) {
+func (registry *DestinationRegistry) GetImageReference(repo, reference string) (ref.Ref, error) {
 	return registry.tempStorage.GetImageReference(repo, reference)
 }
 
 // finalize a syncing image.
-func (registry *DestinationRegistry) CommitImage(imageReference types.ImageReference, repo, reference string) error {
-	imageStore := registry.storeController.GetImageStore(repo)
-
-	tempImageStore := getImageStoreFromImageReference(imageReference, repo, reference, registry.log)
+func (registry *DestinationRegistry) CommitAll(repo string, imageReference ref.Ref) error {
+	tempImageStore := getImageStoreFromImageReference(repo, imageReference, registry.log)
 
 	defer os.RemoveAll(tempImageStore.RootDir())
 
-	registry.log.Info().Str("syncTempDir", path.Join(tempImageStore.RootDir(), repo)).Str("reference", reference).
+	registry.log.Info().Str("syncTempDir", path.Join(tempImageStore.RootDir(), repo)).Str("repository", repo).
 		Msg("pushing synced local image to local registry")
 
-	var lockLatency time.Time
-
-	manifestBlob, manifestDigest, mediaType, err := tempImageStore.GetImageManifest(repo, reference)
+	index, err := storageCommon.GetIndex(tempImageStore, repo, registry.log)
 	if err != nil {
 		registry.log.Error().Str("errorType", common.TypeOf(err)).
-			Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).Str("reference", reference).
-			Msg("couldn't find synced manifest in temporary sync dir")
+			Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).
+			Msg("failed to get repo index from temp sync dir")
 
 		return err
 	}
 
-	// is image manifest
-	switch mediaType {
-	case ispec.MediaTypeImageManifest:
-		if err := registry.copyManifest(repo, manifestBlob, reference, tempImageStore); err != nil {
+	seen := []godigest.Digest{}
+
+	for _, desc := range index.Manifests {
+		reference := GetDescriptorReference(desc)
+
+		if err := registry.copyManifest(repo, desc, reference, tempImageStore, seen); err != nil {
 			if errors.Is(err, zerr.ErrImageLintAnnotations) {
 				registry.log.Error().Str("errorType", common.TypeOf(err)).
-					Err(err).Msg("couldn't upload manifest because of missing annotations")
+					Err(err).Msg("failed to upload manifest because of missing annotations")
 
 				return nil
 			}
 
 			return err
 		}
-	case ispec.MediaTypeImageIndex:
-		// is image index
-		var indexManifest ispec.Index
+	}
+
+	return nil
+}
+
+func (registry *DestinationRegistry) CleanupImage(imageReference ref.Ref, repo string) error {
+	var err error
+
+	dir := strings.TrimSuffix(imageReference.Path, repo)
+	if _, err = os.Stat(dir); err == nil {
+		if err := os.RemoveAll(strings.TrimSuffix(imageReference.Path, repo)); err != nil {
```

Comment:
Can we get coverage for this? It seems we don't have tests where os.Stat() succeeds above.